pkg/backends/calico/routes.go

```diff
@@ -15,30 +15,240 @@ package calico
+		log.Warn("setRouteForSvc: both service and endpoint cannot be nil, passing...")
+		return
+	}
+
+	var (
+		key string
+		err error
+	)
+
+	if svc == nil {
+		// ep received but svc nil
+		// get key
+		key, err = cache.MetaNamespaceKeyFunc(ep)
+		if err != nil {
+			log.WithField("svc", svc.Name).WithError(err).Warn("setRouteForSvc: error on retrieving key for endpoint, passing")
+			return
+		}
+		// get svc
+		svcIface, exists, err := rg.svcIndexer.GetByKey(key)
+		if err != nil {
+			log.WithField("key", key).WithError(err).Warn("setRouteForSvc: error on retrieving service for key, passing")
+			return
+		} else if !exists {
+			log.WithField("key", key).Warn("setRouteForSvc: service for key not found, passing")
+			return
+		}
+		svc = svcIface.(*v1.Service)
+	} else if ep == nil {
+		// svc received but ep nil
+		// get key
+		key, err = cache.MetaNamespaceKeyFunc(svc)
+		if err != nil {
+			log.WithField("ep", ep.Name).WithError(err).Warn("setRouteForSvc: error on retrieving key for service, passing")
+			return
+		}
+		// get ep
+		epIface, exists, err := rg.epIndexer.GetByKey(key)
+		if err != nil {
+			log.WithField("key", key).WithError(err).Warn("setRouteForSvc: error on retrieving endpoint for key, passing")
+			return
+		} else if !exists {
+			log.WithField("key", key).Warn("setRouteForSvc: service for endpoint not found, passing")
+			return
+		}
+		ep = epIface.(*v1.Endpoints)
+	}
+
+	// always set if externalTrafficPolicy == cluster
+	if svc.Spec.ExternalTrafficPolicy == v1.ServiceExternalTrafficPolicyTypeCluster {
+		rg.svcRouteMap[key] = svc.Spec.ClusterIP
+		return
+	}
+
+	// add svc clusterIP if node contains at least one endpoint for svc
+	found := false
+	for _, subset := range ep.Subsets {
+		for _, address := range subset.Addresses {
+			if address.NodeName == nil || *address.NodeName != rg.nodeName {
+				continue
+			}
+			rg.svcRouteMap[key] = svc.Spec.ClusterIP
+			found = true
+			break
 		}
 	}
+
+	// remove the route if in svcRouteMap and no valid endpoints were found
+	if _, exists := rg.svcRouteMap[key]; !found && exists {
+		delete(rg.svcRouteMap, key)
+	}
+
+	rg.updateRoutes()
+	log.Debug("-setRouteForSvc")
+}
+
+// unsetRouteForSvc removes the route from the svcRouteMap
+// but checks to see if it wasn't already deleted by its sibling resource
+func (rg *routeGenerator) unsetRouteForSvc(obj interface{}) {
+	log.Debug("+unsetRouteForSvc")
+	// generate key
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		log.WithError(err).Warn("unsetRouteForSvc: error on retrieving key for object, passing")
+		return
+	}
+
+	// mutex
+	rg.Lock()
+	defer rg.Unlock()
+
+	// pass if already deleted
+	if _, exists := rg.svcRouteMap[key]; !exists {
+		return
+	}
+
+	// delete
+	delete(rg.svcRouteMap, key)
+	rg.updateRoutes()
+	log.Debug("-unsetRouteForSvc")
+}
+
+// onSvcAdd is called when a k8s service is created
+func (rg *routeGenerator) onSvcAdd(obj interface{}) {
+	rg.setRouteForSvc(obj.(*v1.Service), nil)
```

Comment:
Probably will always be a Service object, but best to gracefully handle this case rather than panic, else we'll bring the whole process down. Same below for the other cast operations.